backend/internal/system/crypto/hash/hash.go

```diff
@@ -20,40 +20,132 @@
+// based on the reference credential's algorithm.
+func Verify(credentialValueToVerify []byte, referenceCredential Credential) bool {
+	logger := log.GetLogger().With(log.String(log.LoggerKeyComponentName, "HashProvider"))
+	algorithm := referenceCredential.Algorithm
+
+	switch algorithm {
+	case sha256Algorithm:
+		logger.Debug("Using SHA256 hash provider for verification as per credential algorithm.")
+		return verifySHA256Credential(credentialValueToVerify, referenceCredential)
+	case pbkdf2Algorithm:
+		logger.Debug("Using PBKDF2 hash provider for verification as per credential algorithm.")
+		return verifyPBKDF2Credential(credentialValueToVerify, referenceCredential)
+	default:
+		logger.Error("Unsupported hash algorithm in credential", log.String("algorithm", referenceCredential.Algorithm))
+		return false
+	}
 }
 
-// HashString returns a SHA-256 hash of the input string.
-func HashString(input string) string {
-	return Hash([]byte(input))
+// newSHA256Credential generates a SHA256 hash of the input data combined with salt.
+func newSHA256Credential(credentialValue []byte) Credential {
+	credSalt, _ := generateSalt()
+	credentialValue = append(credentialValue, credSalt...)
+	hash := sha256.Sum256(credentialValue)
+
+	return Credential{
+		Algorithm: sha256Algorithm,
+		Hash:      hex.EncodeToString(hash[:]),
+		Salt:      hex.EncodeToString(credSalt),
+	}
 }
 
-// HashStringWithSalt hashes the input string with the given salt and returns the hex-encoded hash.
-func HashStringWithSalt(input, salt string) (string, error) {
-	hasher := sha256.New()
-	_, err := hasher.Write([]byte(input + salt))
+// verifySHA256Credential checks if the SHA256 hash of the input data and salt matches the expected hash.
+func verifySHA256Credential(credentialValueToVerify []byte, referenceCredential Credential) bool {
+	saltBytes, err := hex.DecodeString(referenceCredential.Salt)
 	if err != nil {
-		return "", err
+		return false
 	}
-	return hex.EncodeToString(hasher.Sum(nil)), nil
+	credentialValueToVerify = append(credentialValueToVerify, saltBytes...)
+	hashedData := sha256.Sum256(credentialValueToVerify)
+	return referenceCredential.Hash == hex.EncodeToString(hashedData[:])
 }
 
-// GenerateSalt generates a random salt string.
-func GenerateSalt() (string, error) {
+// newPBKDF2Credential generates a PBKDF2 hash of the input data using the provided salt.
+func newPBKDF2Credential(credentialValue []byte) Credential {
+	credSalt, _ := generateSalt()
+	logger := log.GetLogger().With(log.String(log.LoggerKeyComponentName, "PBKDF2HashProvider"))
+	hash, err := pbkdf2.Key(sha256.New, string(credentialValue), credSalt, defaultPBKDF2Iterations, defaultPBKDF2KeyLength)
+	if err != nil {
+		logger.Error("Error hashing data with PBKDF2: %v", log.Error(err))
+		return Credential{}
+	}
+	return Credential{
+		Algorithm: pbkdf2Algorithm,
+		Hash:      hex.EncodeToString(hash),
+		Salt:      hex.EncodeToString(credSalt),
+	}
```

Comment:
Since we need the credential to be self contained, don't we need `defaultPBKDF2Iterations`, `defaultPBKDF2KeyLength` etc to embedded into the Crednetial object?